### DISCUSSION TITLE
Is there a way to map an out param1** that is an array of structs of length param2?

### DISCUSSION THREAD

**Discussion by kmgallahan** [SharpGenTools/SharpGenTools#201] [Q&A] [answered]
As per the title, I can't quite figure out if it is possible to map [this type of function](https://github.com/kmgallahan/AllSystemsGo.SonyCameraRemoteSDK_SGT/blob/6d89c003ea303c53b31a8df127276c60ced1f3d6/AllSystemsGo.SonyCameraRemoteSDK/CrSDK/CameraRemote_SDK.h#L81):

```
CrError GetDeviceProperties(/*in*/ CrDeviceHandle deviceHandle, /*out*/CrDeviceProperty** properties, /*out*/ CrInt32* numOfPropoties);
```

Where the second `out` param is a `**` to an array of structs of type CrDeviceProperty of the length of the 3rd param, as per the documentation:

> The first one is the device handle that specifies the device, the second 
one is the pointer to CrDeviceProperty pointer that receives the acquired property list, and the third one 
receives the size of the CrDeviceProperty list.

I'm aware of the [Length Relation](https://sharpgentools.readthedocs.io/en/latest/relations.html#length-relation), but it appears to only be for `in` params. I also don't know how to deal with `**` params.

[Current mapping for this header](https://github.com/kmgallahan/AllSystemsGo.SonyCameraRemoteSDK_SGT/blob/master/AllSystemsGo.SonyCameraRemoteSDK/Mapping.xml) produces:

```
public static unsafe int GetDeviceProperties(long deviceHandle, out System.IntPtr properties, out int numOfProperties)
{
      int __result__;
      fixed (void *numOfProperties_ = &numOfProperties)
          fixed (void *properties_ = &properties)
              __result__ = GetDeviceProperties_(deviceHandle, properties_, numOfProperties_);
      return __result__;
}
```

**Comment by andrew-boyarshin** [comment]
I believe we've blocked this specifically. It's due to the unknown allocation strategy, we cannot guarantee how it's supposed to be deallocated, therefore we can't lose the raw pointer. We offer `MarshallingHelpers.Read` for that.

**Comment by kmgallahan** [comment]
Did you mean `MarshallingHelpers.FromPointer`? I don't see a read anywhere.

https://github.com/SharpGenTools/SharpGenTools/blob/fe286fc05db168114da936b641034e074bbc10ca/SharpGen.Runtime/MarshallingHelpers.cs#L14

**Comment by andrew-boyarshin** [comment, marked as answer]
Whoops, I meant `MemoryHelpers`, actually.
<https://github.com/SharpGenTools/SharpGenTools/blob/fe286fc05db168114da936b641034e074bbc10ca/SharpGen.Runtime/MemoryHelpers.cs#L121-L194>

**Reply by kmgallahan** [reply]
Great success! Getting this function to work was a prerequisite to executing the shutter release command which just allowed me to take my first photo! Thank you! ❤❤❤

Is there any way to reward you for your kind efforts in helping me out? The project has no donation methods that I see.

**Comment by kmgallahan** [comment]
@jkoritzinsky 

Are there plans to add a "hint" to SharpGenTools to let it know how to deallocate this type of pointer? At least for this use case I just have boilerplate:

```
var getDevicePropertiesResult = CR_Core.GetDeviceProperties(handle, out var props, out var numProps);
var deviceProps = new CrDeviceProperty[numProps];
MemoryHelpers.Read(props, deviceProps, 0, numProps);
```

Perhaps even an overload of `MemoryHelpers.Read` for convenience:

```
public static IntPtr Read<T>(IntPtr source, out T[] data, int count) where T : unmanaged
{
    return Read(source, new ReadOnlySpan<T>(data = new T[count]), count);
}
```

To simply down to:

```
var getDevicePropertiesResult = CR_Core.GetDeviceProperties(handle, out var props, out var numProps);
MemoryHelpers.Read(props, out CrDeviceProperty[] deviceProps, numProps);
```
